sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1753,15 +1753,26 @@ class DataFrame private[sql](
    * Converts a JavaRDD to a PythonRDD.
    */
   protected[sql] def javaToPython: JavaRDD[Array[Byte]] = {
-    val structType = schema  // capture it for closure
-    val rdd = queryExecution.toRdd.map(EvaluatePython.toJava(_, structType))
-    EvaluatePython.javaToPython(rdd)
+    if (isOutputPickled) {
+      queryExecution.toRdd.map(_.getBinary(0))
```

Comment:
It will be faster when do the count() in Python, then you don't need to pass all these rows into JVM.

Even for groupByKey(), if the key_func is Python function, the rows will be deserialized in Python.

For in-memory cache, the rows are packed as batched in JVM, so it's not true that we need the rows are serialized separately to be able to be processed in JVM.